File: src/collectors/ssh.ts
import { join } from "node:path";
import { REDACTION_MARKER } from "../utils/constants";
import type { Collector, CollectorResult } from "./types";
import { makeSection } from "./types";

interface SshHost {
  host: string;
  hostname: string;
  identityFile: string;
}

function parseSshConfig(content: string): SshHost[] {
  const hosts: SshHost[] = [];
  let current: Partial<SshHost> | null = null;

  for (const line of content.split("\n")) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) continue;

    const hostMatch = trimmed.match(/^Host\s+(.+)/i);
    if (hostMatch) {
      if (current?.host) hosts.push(current as SshHost);
      current = { host: hostMatch[1].trim(), hostname: "", identityFile: "" };
      continue;
    }

    if (!current) continue;

    const hostnameMatch = trimmed.match(/^HostName\s+(.+)/i);
    if (hostnameMatch) current.hostname = hostnameMatch[1].trim();

    const identityMatch = trimmed.match(/^IdentityFile\s+(.+)/i);
    if (identityMatch) current.identityFile = identityMatch[1].trim();
  }

  if (current?.host) hosts.push(current as SshHost);
  return hosts;
}

export const collectSsh: Collector = async (ctx) => {
  const result: CollectorResult = {};
  const configPath = join(ctx.home, ".ssh/config");
  const file = Bun.file(configPath);

  if (!(await file.exists())) return result;

  const content = await file.text();
  const hosts = parseSshConfig(content);

  if (!hosts.length) return result;

  const items = hosts.map((h) => {
    const hn = ctx.redact ? REDACTION_MARKER : h.hostname;
    const id = ctx.redact ? REDACTION_MARKER : h.identityFile;
    const raw = `${h.host} | ${hn} | ${id}`;
    return { raw, columns: [h.host, hn, id] };
  });

  result["ssh.hosts"] = makeSection("ssh.hosts", { items });
  return result;
};

File: internal/collect/ssh.go
package collect

import (
	"path/filepath"
	"regexp"
	"strings"

	"github.com/doguyilmaz/dothaven/internal/snapshot"
)

const sshRedactionMarker = "[REDACTED]"

var (
	sshHostRe     = regexp.MustCompile(`(?i)^Host\s+(.+)`)
	sshHostNameRe = regexp.MustCompile(`(?i)^HostName\s+(.+)`)
	sshIdentityRe = regexp.MustCompile(`(?i)^IdentityFile\s+(.+)`)
)

// SSHHost is a single parsed entry from ~/.ssh/config.
type SSHHost struct {
	Host         string
	HostName     string
	IdentityFile string
}

// ParseSSHConfig parses an ssh config into ordered host entries. A new entry
// begins at each Host line; HostName/IdentityFile attach to the current entry.
func ParseSSHConfig(content string) []SSHHost {
	var hosts []SSHHost
	var current *SSHHost

	for _, line := range strings.Split(content, "\n") {
		trimmed := strings.TrimSpace(line)
		if trimmed == "" || strings.HasPrefix(trimmed, "#") {
			continue
		}

		if m := sshHostRe.FindStringSubmatch(trimmed); m != nil {
			if current != nil && current.Host != "" {
				hosts = append(hosts, *current)
			}
			current = &SSHHost{Host: strings.TrimSpace(m[1])}
			continue
		}

		if current == nil {
			continue
		}

		if m := sshHostNameRe.FindStringSubmatch(trimmed); m != nil {
			current.HostName = strings.TrimSpace(m[1])
		}

		if m := sshIdentityRe.FindStringSubmatch(trimmed); m != nil {
			current.IdentityFile = strings.TrimSpace(m[1])
		}
	}

	if current != nil && current.Host != "" {
		hosts = append(hosts, *current)
	}
	return hosts
}

// SSHCollector reads ~/.ssh/config and emits the "ssh.hosts" section with items
// columned [host, hostname, identity]. HostName and IdentityFile are redacted
// when c.Redact is set.
func SSHCollector(c Ctx) snapshot.Snapshot {
	out := snapshot.Snapshot{}

	configPath := filepath.Join(c.Home, ".ssh", "config")
	data, err := c.Env.ReadFile(configPath)
	if err != nil {
		return out
	}

	hosts := ParseSSHConfig(string(data))
	if len(hosts) == 0 {
		return out
	}

	items := make([]snapshot.Item, len(hosts))
	for i, h := range hosts {
		hn := h.HostName
		id := h.IdentityFile
		if c.Redact {
			hn = sshRedactionMarker
			id = sshRedactionMarker
		}
		items[i] = snapshot.Item{
			Raw:     h.Host + " | " + hn + " | " + id,
			Columns: []string{h.Host, hn, id},
		}
	}

	out["ssh.hosts"] = snapshot.Section{Items: items}
	return out
}
